papas/aliceproto/DAG.py

```diff
@@ -0,0 +1,232 @@
+             
+    def bfs_recursive(self,nodes, linktype ):
+        '''Breadth first recursive implementation
+        each recursion is one level down the tree
+        linktype can be "children", "parents","undirected" '''
+        linknodes=[]
+        if len(nodes) is 0:
+            return 
+        
+        for node in nodes: # collect a list of all the next level of nodes
+            if (self.visited.get(node, False)):  
+                continue              
+            linknodes.extend(node.get_linked_nodes(linktype))        
+        for node in nodes: #add these nodes onto list and mark as visited
+            if (self.visited.get(node, False)):  
+                continue            
+            node.accept(self)
+        
+        self.bfs_recursive(linknodes,  linktype)
+
+          
+class BreadthFirstSearchIterative(object):
+    
+    def __init__(self,root, linktype):
+        '''Perform the breadth first iterative search of the nodes'''
+        self.visited={}
+        self.result=[]
+        self.bfs_iterative(root,linktype)       
+
+    def visit(self, node):
+        if self.visited.get(node, False):
+            return
+        self.result.append( node )
+        self.visited[node]=True
+        
+
+        
+    def bfs_iterative(self,node, linktype ):
+        '''Breadth first iterative implementation
+        using a deque to order the nodes 
+        linktype can be "children", "parents","undirected" '''
+            
+        # Create a deque for the Breadth First Search
+        todo = deque()
+        todo.append( node)
+        
+        while len(todo):
+            node = todo.popleft()
+            if self.visited.get(node,False): #check if already processed
+                continue
+            node.accept(self)
+            for linknode in node.get_linked_nodes(linktype):
+                if self.visited.get(linknode,False): #check if already processed
+                    continue
+                todo.append( linknode)  
+                
+                
+class DAGFloodfill(object):
+    
+    def __init__(self,elements,first_label=1):
+        '''Use Breadth first search to find connected groups'''
+        self.visited={}
+        self.label = first_label
+        self.visited = dict()
+        self.blocks = []
+        for uid, node in elements.iteritems():
+                if self.visited.get(node, False): #already done so skip the rest
+                    continue
+                
+                #find connected nodes
+                bfs=BreadthFirstSearchIterative(node,"undirected")
+                
+                # set all connected elements to have a visited flag =true
+                for n in bfs.result :
+                    self.visited.update({n: True})
+                #add into the set of blocks
+                self.blocks.append( bfs.result)
+                self.label += 1
+        
+        
+      
+            
+
+class TreeTestCase( unittest.TestCase ):
```

Comment:
test cases should be in separate files next to the module they test, see the other tests in this package